aip/general/0236.md

```diff
@@ -0,0 +1,355 @@
+
+- /compute/v1/locations/global/policies/{policy_id}
+
+The experimental versions of the resource used for preview mode are stored as a nested collection underneath the policy. The resource type will follow the convention "<resource_name> + Experiment”.
+
+- /compute/v1/locations/global/policies/{policy_id}/policyExperiments/{policy_experiment_id}
+
+**Note if the resource is called OrgPolicy, this collection would be called OrgPolicyExperiments.
+
+This nested collection is extensible to progressive rollout and draft policy use-cases as well, which are out of scope for this AIP.
+
+The following proto represents an experiment Policy used for preview mode. It **must** contain the required top-level fields like name,  a nested message representing the parent-level Policy fields, and the necessary Metadata fields for preview mode.
+
+```proto
+message PolicyExperiment {
+
+  // The name of the PolicyExperiment.
+  string name = 1;
+ 
+  // System generated UUID whenever the resource is updated.
+  string etag = 2;
+
+  // The fields used for log evaluation and eventual promotion.
+  Policy policy = 3;
+
+  // Metadata for preview mode only.
+  PreviewMetadata preview_metadata = 4 [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+- A new proto **should** be used for PolicyExperiment with the Policy resource proto inlined inside.
+The inlined Policy field contains any changes to the parent resource which are to be tested using preview mode.
+- The name inside the embedded Policy message **must** be the name of the parent. 
+- When the user is ready to promote an experimental configuration, the policy message field **must** be copied into the live resource and the experiment is deleted. This can be done manually or via a custom verb for commit described below.
+- A product **may** support multiple new configurations concurrently in preview mode for a singular parent policy. This is represented using two or more resources in the nested collection.
+  - The number of configurations in preview mode for a given parent **may** be capped at a certain number.
+  - 403 **may** be returned if the number of supported actively previewed policies for a given parent are exceeded.
+- All resources under the experiments collection **must** be cascade-deleted if the top-level resource is deleted.
+- The etag on the Policy Experiment **should** guidance on Resource freshness validation ([AIP-154][aip-154]) and will be a randomly generated UUID ([AIP-148][aip-148]). The etag **should** be present in the preview logs.
+  - The parent resource **should** have an etag as well.
+  - Both etags **should** be present in the preview logs to uniquely identify the version of the live and preview policies. These will change when either policy configurations change.
+
+[aip-148]: https://aip.dev/148
+[aip-154]: https://aip.dev/154
+
+### Metadata
+A Metadata object is used to track all variables pertinent to a preview.
+ 
+```proto 
+message PreviewMetadata {
```

Comment:
Considering that PolicyExperiment is a resource --- these fields all seem consistent with state and *_time fields you would expect to find on the resource itself. Consider moving up to top-level output_only fields?

As top-level fields of experiment resource, start_time/stop_time would be consistent with other *_time fields you find in AIP-148. Along those lines, and create_time/update_time would be something to consider including on experiment resource as well. 

And the state as the top-level seems like it would be consistent with AIP-216 field on the experiment.

Reply:
So...this is to bunch together fields separately for Preview vs Progressive Rollout. So we want this separation for our internal use-cases.

Since this is public documentation that won't be covering progressive rollout, what do you recommend is the right approach? I could go with top-level fields as the default but recommend grouping if PolicyExperiment is needed for other use-cases.

Reply:
That's a good question, two thoughts there - just because the nature of the nested preview fields (the status and method timestamps) really leave a sense this may be compacting two resources into one...

* thought one

A quick change to get those fields top-level would be to align the field names with the custom method names

startPreview -> start_preview_time
stopPreview -> stop_preview_time
enum preview_state

Of course having a prefixed status like that raises the question: what would the "Experiment" resource's { enum state } be on its own - and would the preview states just be members of that? Or are the other non-preview scenarios mutually exclusive set of states - and those may occur simultaneously?

* thought two

Another avenue that this resource-ness of preview metadata could lead you down is to think that a preview is a noun that exists under an experiment.

`policies/p-1/experiments/e-1/previews/pre-1`

A bit like the "job object" pattern in that way. You create a new experiment, then you create a new preview, and the preview's create_time is when it started. The permission to create a preview is granted to fewer people than the permission to create an experiment (same people who would have startPreview custom method access).

The preview is in state active until cancel or delete is called... You can also preview and re-preview without overwriting/destroying preview history (assuming you cancel, or have a duration field, rather than delete to stop)

* thought three

Keep as-is and adjust later if needed. The proposal has already been reviewed by several people, and there's no reason to block on progress a few fields.

Reply:
Thanks for your thoughts Lou. 

#1 leads to worse readability when extending to other use-cases IMO. The use of nested messages for each  safe rollout use-case metadata helps with separation of use-cases specific variables that we are feeling is an asset for elegance.

#2 I think this is starting to get into "too-many-resources" land but I'm open to adjusting this as suggested in #3 down the line, but not blocking for now.